lib/puppet/transaction.rb

```diff
@@ -148,13 +148,33 @@ def evaluate(&block)
       persistence.save if catalog.host_config?
     end
 
+    # Graph cycles are returned as an array of arrays
+    # - outer array is an array of cycles
+    # - each inner array is an array of resources involved in a cycle
+    # Short circuit resource evaluation if we detect cycle(s) in the graph. Mark
+    # each corresponding resource as failed in the report before we fail to
+    # ensure accurate reporting.
+    graph_cycle_handler = lambda do |cycles|
+      cycles.each do |cycle|
+        cycle.each do |resource|
+          # We add a failed resource event to the status to ensure accurate
+          # reporting through the event manager. Adding an event with status of
+          # 'failure' causes the status itself to be failed.
+          event = resource.event(:name => :resource_error, :status => "failure", :message => _('resource is part of a dependency cycle'))
+          resource_status(resource).add_event(event)
```

Comment:
So I tried to create a test repro scenario where we have a resource in more than one cycle as detected by simple_graph, and I couldn't. At first, I thought this must be a bug.. i.e., given this graph:
a => b, a => c
b => a
c => a
I was expecting two cycles reported, with _a_ in each of them, but instead one cycle was reported, containing, _a_, _b_, and _c_:
![foo](https://user-images.githubusercontent.com/819133/26903495-536a05e4-4b92-11e7-89ca-38751dc827a9.png)

Turns out, its mathematically impossible for a vertex to be in more than one strongly connected component (ie what we consider a resource cycle) - rather,  the whole thing is the strongly connected component. https://math.stackexchange.com/questions/358094/how-can-it-be-proved-that-each-vertex-can-be-at-most-in-one-strongly-connected-c.

[This test](https://github.com/puppetlabs/puppet/blob/master/spec/unit/graph/simple_graph_spec.rb#L381-L386) seems to agree. Did you find differently?

I'm still happy to flatten the arrays but I'm no longer sure it's going to have any effect ultimately.